File: lab3-333/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <libgen.h>
#include <sys/stat.h>
#include <sys/types.h>

#define PATH_MAX 1024
#define BUFSIZ 8192

// eprintf and exit wrapper
void fatal(const char *msg) {
    fprintf(stderr, "Error: %s\n", msg);
    exit(1);
}

// Check if path is a file
int is_file(const char *path) {
    struct stat st;
    if (stat(path, &st) == -1) return 0;
    return S_ISREG(st.st_mode);
}

// Check if path is a directory
int is_dir(const char *path) {
    struct stat st;
    if (stat(path, &st) == -1) return 0;
    return S_ISDIR(st.st_mode);
}

// Convert to absolute path
// TODO: a better solution
char *to_abs_path(const char *path, char *resolved_path) {
    if (path[0] != '/') {
        // relative path
        char cwd[1024];
        if (getcwd(cwd, sizeof(cwd)) != NULL) {
            snprintf(resolved_path, 1028, "%s/%s", cwd, path);
        } else {
            perror("getcwd");
            exit(1);
        }
    } else {
        // absolute path
        strncpy(resolved_path, path, 1024);
    }
    return resolved_path;
}

// Get file name from path
const char *get_file_name(const char *path) {
    const char *slash = strrchr(path, '/');
    return slash ? slash + 1 : path;
}

// Copy file from src to dest
void copy_file(const char *src, const char *dest) {
    char abs_src[PATH_MAX], abs_dest[PATH_MAX];
    to_abs_path(src, abs_src);
    to_abs_path(dest, abs_dest);

    if (!is_file(src)) {
        fprintf(stderr, "File %s does not exist\n", src);
        exit(1);
    }

    if (strcmp(abs_src, abs_dest) == 0) {
        fprintf(stderr, "File %s is tried to be copied to itself\n", src);
        exit(1);
    }

    if (is_file(dest)) {
        printf("File %s already exists. Overwrite? [y/n] ", src);
        char ch = getchar();
        if (ch != 'y' && ch != 'Y') {
            printf("Skipping file %s.\n", src);
            return;
        }
        while (getchar() != '\n'); // flush stdin
    }

    int in_fd = open(src, O_RDONLY);
    if (in_fd < 0) {
        perror("open src");
        exit(1);
    }

    int out_fd = open(dest, O_WRONLY | O_CREAT | O_TRUNC, 0644);
    if (out_fd < 0) {
        perror("open dest");
        close(in_fd);
        exit(1);
    }

    // Buffered copy loop
    char buf[BUFSIZ];
    ssize_t bytes;
    while ((bytes = read(in_fd, buf, BUFSIZ)) > 0) {
        if (write(out_fd, buf, bytes) != bytes) {
            perror("write");
            close(in_fd);
            close(out_fd);
            exit(1);
        }
    }

    if (bytes < 0) {
        perror("read");
    }

    close(in_fd);
    close(out_fd);
}

// Copy multiple files to a destination directory
void copy_files(int file_count, char **files, const char *destination) {
    if (!is_dir(destination)) {
        fatal("Destination is not a directory");
    }

    for (int i = 0; i < file_count; i++) {
        const char *src = files[i];
        const char *filename = get_file_name(src);
        char dest_path[PATH_MAX];
        snprintf(dest_path, sizeof(dest_path), "%s/%s", destination, filename);
        copy_file(src, dest_path);
    }
    printf("Files successfully copied to %s\n", destination);
}

// Move multiple files to a destination directory
void move_files(int file_count, char **files, const char *destination) {
    if (!is_dir(destination)) {
        fatal("Destination is not a directory");
    }

    for (int i = 0; i < file_count; i++) {
        const char *src = files[i];
        const char *filename = get_file_name(src);
        char dest_path[PATH_MAX];
        snprintf(dest_path, sizeof(dest_path), "%s/%s", destination, filename);

        // Try hard link first
        if (link(src, dest_path) != 0) {
            perror("link");
            printf("Copying %s instead.\n", src);
            copy_file(src, dest_path);
        }

        // Remove original
        if (unlink(src) != 0) {
            perror("unlink");
            exit(1);
        }
    }
    printf("Files successfully moved to %s\n", destination);
}

// Entry point
int main(int argc, char *argv[]) {
    if (argc < 3) {
        fprintf(stderr, "Usage: %s source1 [source2 ...] destination\n", argv[0]);
        exit(1);
    }

    const char *progname = get_file_name(argv[0]);

    if (strstr(progname, "copy")) {
        copy_files(argc - 2, &argv[1], argv[argc - 1]);
    } else if (strstr(progname, "move")) {
        move_files(argc - 2, &argv[1], argv[argc - 1]);
    } else {
        fatal("Invalid program name, must contain 'copy' or 'move'");
    }

    return 0;
}

File: lab3-333/src/main.rs
// in c code, all `anyhow` errors should be simply eprintf and exit(1)

use std::{
    env, fs,
    io::{self, Read, Write},
    path::{Path, PathBuf},
};

use anyhow::{anyhow, Context, Ok, Result};

const BUFSIZ: usize = 8192; // in c code we have this macro, include it

fn copy_file(src: &Path, dest: &Path) -> anyhow::Result<()> {
    dbg!(src, dest);
    // check if file exists
    // in c code, is_file and is_dir should be encapsulated in functions
    if !fs::metadata(src)?.is_file() {
        return Err(anyhow!("File {} does not exist", src.display()));
    }
    // check if the file is tried to be copied to itself
    // in c code, you need to encapsulate a function "to_abs_path"
    // and check if the abs paths are the same
    if src == dest {
        return Err(anyhow!(
            "File {} is tried to be copied to itself",
            src.display()
        ));
    }
    // check if file already exists in destination
    if dest.is_file() {
        println!(
            "File {} already exists in destination. Overwrite? [y/n] ",
            src.display()
        );
        let mut input = String::new();
        io::stdin().read_line(&mut input)?;
        if input.trim() != "y" {
            println!("Skipping file {}.", src.display());
            return Ok(());
        }
    }
    // in c code, file_name should be encapsulated in a function
    let mut src_file = fs::File::open(src)?;
    let mut dest_file = fs::File::create(dest)?;
    let mut dest_writer = fs::OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(true)
        .open(dest)?;
    let mut buffer = [0u8; BUFSIZ];
    loop {
        let bytes_read = src_file.read(&mut buffer)?;
        if bytes_read == 0 {
            break;
        }
        dest_file.write_all(&buffer[..bytes_read])?;
    }
    dest_writer.flush()?;
    Ok(())
}

fn copy_files(files: &[String], destination: &String) -> anyhow::Result<()> {
    dbg!(files, destination);
    // i'm not sure if c code has this kind of abstraction like `PathBuf`,
    // if no, choose another way to implement this
    let files = files.to_vec().iter().map(PathBuf::from).collect::<Vec<_>>();
    // check if destination is a directory
    if !(fs::metadata(destination)?.is_dir()) {
        return Err(anyhow!("Destination is not a directory"));
    }
    fs::create_dir_all(destination)
        .with_context(|| format!("Failed to create destination directory: {}", destination))?;
    for file in &files {
        copy_file(
            file,
            &Path::new(destination).join(file.file_name().unwrap_or_default()),
        )?;
    }
    println!(
        "{} successfully copied to {}",
        files
            .iter()
            .map(|f| f.display().to_string())
            .collect::<Vec<_>>()
            .join(" "),
        destination,
    );
    Ok(())
}

fn move_files(files: &[String], destination: &String) -> anyhow::Result<()> {
    dbg!(files, destination);
    let files = files.to_vec().iter().map(PathBuf::from).collect::<Vec<_>>();
    // check if destination is a directory
    if !(fs::metadata(destination)?.is_dir()) {
        return Err(anyhow!("Destination is not a directory"));
    }
    fs::create_dir_all(destination)
        .with_context(|| format!("Failed to create destination directory: {}", destination))?;
    for file in &files {
        let dest_path = Path::new(destination).join(file.file_name().unwrap_or_default());
        // first try: hard link
        // in c code you might need link()
        println!("Creating hard link for {}...", file.display());
        if let Err(e) = fs::hard_link(file, &dest_path) {
            println!(
                "Failed to create hard link for {}: {}\nCopy and remove original file instead.",
                file.display(),
                e
            );
            // if hard link fails, try copy
            copy_file(file, &dest_path)?;
            // then remove the original file
            // in c code use unlink()
            fs::remove_file(file)?;
        }
    }
    println!(
        "{} successfully moved to {}",
        files
            .iter()
            .map(|f| f.display().to_string())
            .collect::<Vec<_>>()
            .join(" "),
        destination,
    );
    Ok(())
}

fn main() -> Result<()> {
    let args = env::args().collect::<Vec<_>>();
    if args[0].contains("copy") {
        if args.len() < 3 {
            println!("Usage: copy source1 [source2 ...] destination");
        } else {
            copy_files(&args[1..args.len() - 1], &args[args.len() - 1])?;
        }
    } else if args[0].contains("move") {
        if args.len() < 3 {
            println!("Usage: move source1 [source2 ...] destination");
        } else {
            move_files(&args[1..args.len() - 1], &args[args.len() - 1])?;
        }
    } else {
        panic!("Unreachable code here!");
    }
    Ok(())
}
